internal/http/services/appprovider/appprovider.go

```diff
@@ -135,34 +126,87 @@ func (s *svc) handleNew(w http.ResponseWriter, r *http.Request) {
 
 	client, err := pool.GetGatewayServiceClient(s.conf.GatewaySvc)
 	if err != nil {
-		ocmd.WriteError(w, r, ocmd.APIErrorServerError, "error getting grpc gateway client", err)
+		writeError(w, r, appErrorServerError, "error getting grpc gateway client", err)
 		return
 	}
 
 	if r.URL.Query().Get("template") != "" {
 		// TODO in the future we want to create a file out of the given template
-		ocmd.WriteError(w, r, ocmd.APIErrorInvalidParameter, "Template not implemented",
-			errtypes.NotSupported("Templates are not yet supported"))
+		writeError(w, r, appErrorUnimplemented, "template is not implemented", nil)
+		return
+	}
+
+	parentContainerID := r.URL.Query().Get("parent_container_id")
+	if parentContainerID == "" {
+		writeError(w, r, appErrorInvalidParameter, "missing parent container ID", nil)
+		return
+	}
+
+	parentContainerRef := resourceid.OwnCloudResourceIDUnwrap(parentContainerID)
+	if parentContainerRef == nil {
+		writeError(w, r, appErrorInvalidParameter, "invalid parent container ID", nil)
+		return
+	}
+
+	filename := r.URL.Query().Get("filename")
+	if filename == "" {
+		writeError(w, r, appErrorInvalidParameter, "missing filename", nil)
+		return
+	}
+
+	dirPart, filePart := path.Split(filename)
+	if dirPart != "" || filePart != filename {
+		writeError(w, r, appErrorInvalidParameter, "the filename must not contain a path segment", nil)
+		return
+	}
+
+	statParentContainerReq := &provider.StatRequest{
+		Ref: &provider.Reference{
+			ResourceId: parentContainerRef,
+		},
+	}
+	parentContainer, err := client.Stat(ctx, statParentContainerReq)
+	if err != nil {
+		writeError(w, r, appErrorServerError, "error sending a grpc stat request", err)
+		return
+	}
+
+	if parentContainer.Status.Code != rpc.Code_CODE_OK {
+		writeError(w, r, appErrorNotFound, "the parent container is not accessible or does not exist", err)
 		return
 	}
 
-	target := r.URL.Query().Get("filename")
-	if target == "" {
-		ocmd.WriteError(w, r, ocmd.APIErrorInvalidParameter, "Missing filename",
-			errtypes.UserRequired("Missing filename"))
+	if parentContainer.Info.Type != provider.ResourceType_RESOURCE_TYPE_CONTAINER {
+		writeError(w, r, appErrorInvalidParameter, "the parent container id does not point to a container", nil)
 		return
 	}
 
-	// TODO(lopresti) if target is relative, currently the gateway fails to identify a storage provider (?)
-	// and just returns a CODE_INTERNAL error on InitiateFileUpload.
-	// Therefore for now make sure the target is absolute.
-	if target[0] != '/' {
-		target = "/" + target
+	fileRef := &provider.Reference{
+		ResourceId: parentContainer.Info.Id,
+		Path:       utils.MakeRelativePath(filename),
```

Comment:
Not all drivers support relative references yet. Can we just pass the path as `path.Join(parentContainer.Info.Path, utils.MakeRelativePath(filename)`?

Reply:
Is changed ;-)